# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом.



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [ ]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [ ]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [ ]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [ ]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [ ]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [ ]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [ ]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [ ]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [ ]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [ ]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [ ]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [ ]:
data.duplicated().sum()

71

In [ ]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [ ]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [ ]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [ ]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [ ]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [ ]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [ ]:
# сгруппируем данные по столбцу 'children' с выводом размера группы и количества задолженностей в этой группе
data_grouped_children = data.groupby('children').agg({'debt': ['count', 'sum']})
# добавим столбец процентного значения количества задолженностей по группам
data_grouped_children['percent'] = data_grouped_children['debt']['sum'] / data_grouped_children['debt']['count']
print(data_grouped_children)

           debt         percent
          count   sum          
children                       
0         14091  1063  0.075438
1          4808   444  0.092346
2          2052   194  0.094542
3           330    27  0.081818
4            41     4  0.097561
5             9     0  0.000000


In [ ]:
# Другой способ группировки

data.groupby('children')['debt'].agg(['count','sum','mean'])

count   sum      mean
children                       
0         14091  1063  0.075438
1          4808   444  0.092346
2          2052   194  0.094542
3           330    27  0.081818
4            41     4  0.097561
5             9     0  0.000000

In [ ]:
# Группировка с помощью функции с выводом доли

def my_mean(x): return x.mean()*100

# В кач-ве агрегирования можно любую функцию брать.
data.groupby('children')['debt'].agg(['count', 'sum', my_mean])

count   sum   my_mean
children                       
0         14091  1063  7.543822
1          4808   444  9.234609
2          2052   194  9.454191
3           330    27  8.181818
4            41     4  9.756098
5             9     0  0.000000

In [ ]:
# Группировка с помощью функции с выводом процентов

def my_mean(x): return '{:.2%} '.format(x.mean())

data.groupby('children')['debt'].agg(['count', 'sum', my_mean])

count   sum my_mean
children                     
0         14091  1063  7.54% 
1          4808   444  9.23% 
2          2052   194  9.45% 
3           330    27  8.18% 
4            41     4  9.76% 
5             9     0  0.00%

**Вывод:**
1) Зависимость есть - чем больше детей, тем больше задолженностей в соответствующих группах.
2) Для значения "5 детей" выборка непрезентативная.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [ ]:
# сгруппируем данные по столбцу 'family_status' с выводом размера группы и количества задолженностей в этой группе
data_grouped_family_status = data.groupby('family_status').agg({'debt': ['count', 'sum']})
# добавим столбец процентного значения количества задолженностей по группам
data_grouped_family_status['percent'] = data_grouped_family_status['debt']['sum'] / data_grouped_family_status['debt']['count']
print(data_grouped_family_status)

                        debt        percent
                       count  sum          
family_status                              
Не женат / не замужем   2796  273  0.097639
в разводе               1189   84  0.070648
вдовец / вдова           951   63  0.066246
гражданский брак        4134  385  0.093130
женат / замужем        12261  927  0.075606


In [ ]:
def percent(x): return '{:.2%} '.format(x.mean()) # функция возвращает проценты с двумя знаками после запятой для аргумента.

data.groupby('family_status')['debt'].agg(['count', 'sum', percent]) # здесь группировка по столбцу 'family_status' и
# агрегированный подсчёт трёх значений по столбцу 'debt'. Одно из значений является аргументом для функции my_mean2
# считает среднее значение среди агрегированных
# данных столбца !!!!'debt' !!!!. Если посчитать среднее ряда 0, 1, 0, 0, 0, 1, 1, 1, 0, 1 - то получатся точно проценты 5/10
# или sum/count или 50%

count  sum percent
family_status                            
Не женат / не замужем   2796  273  9.76% 
в разводе               1189   84  7.06% 
вдовец / вдова           951   63  6.62% 
гражданский брак        4134  385  9.31% 
женат / замужем        12261  927  7.56%

**Вывод:**
1) Семейные (7,56%) имеют меньше задолженностей относительно несемейных (9,76%) и живущих в гражданском браке (9,31%)  
2) Вдовцы (6,62%) и в разводе (7,06%) - имеют меньше задолженностей относительно всех групп.

 Овдовевшие это, как правило,  возрастные люди, а они более ответственные. Да и скорее всего у них уже есть взрослые дети, которые могут помогать с выплатами. В целом, можно сказать так: самые безответственные те, кто никогда не был в браке.
</div>


#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [ ]:
# сгруппируем данные по столбцу 'total_income_category' с выводом размера группы и количества задолженностей в этой группе

data.groupby('total_income_category')['debt'].agg(['count', 'sum', percent])

count   sum percent
total_income_category                     
A                         25     2  8.00% 
B                       5014   354  7.06% 
C                      15921  1353  8.50% 
D                        349    21  6.02% 
E                         22     2  9.09%

#### **Вывод:**
1) Больше всего **15921** берут кредитов с уровнем дохода **C**: 50001 <= income <= 200000 - задолженности (8.50%)  
2) Следом **5014** берут кредиты с уровнем дохода **B**: 200001 <= income <= 1000000 - задолженности (7,06%)  
3) Далее **349** берут кредиты с уровнем дохода **D**: 30001 <= income <= 50000 - задолженности (6,02%)  
4) Выборки по доходу **А**: income >= 1000001 (не берут)  и **Е**: 0 <= income <= 30000 (не дают) - непрезентативны  

Прямой зависимости не наблюдается.

Конечно, предложенное заказчиком разбиение не самое понятное. Но похоже, что низкий доход это неблагоприятный фактор. В любом случае, такое распределение вызывает вопросы, поэтому в реальности лучше уточнить у заказчика, все ли в порядке с данными. По таким маленьким группам однозначный вывод делать не стоит, тут я согласна.  

</div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [ ]:
# сгруппируем данные по столбцу 'purpose_category' с выводом размера группы и количества задолженностей в этой группе
data.groupby('purpose_category')['debt'].agg(['count', 'sum', percent])

count  sum percent
purpose_category                            
операции с автомобилем     4279  400  9.35% 
операции с недвижимостью  10751  780  7.26% 
получение образования      3988  369  9.25% 
проведение свадьбы         2313  183  7.91%

###### **Вывод:**
1) Меньше всего задолженностей в группе с недвижимостью (7,26) - первая в пирамиде Маслоу. Самая большая ответственность.  
2) Операции с автомобилем (9,35%) и образованием (9,25%) - менее ответственный подход при взятии кредита  
3) Проведение свадьбы (7,91%) - групповое поручительство скорее всего.  

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

Пропуски найдены в категориях (стаж в днях) days_employed - 2174 и (ежемесячный доход) total_income - 2174.

Наблюдения:
1) Их одновременность говорит о связи событий.

Проверка гипотезы:

Выведем 20 строк с пропущенными значениями

In [ ]:
null_data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')
null_data = null_data[null_data.isnull().any(axis=1)]
print(null_data.shape)
print(null_data.head(20))

(2174, 12)
     children  days_employed  dob_years education  education_id  \
12          0            NaN         65   среднее             1   
26          0            NaN         41   среднее             1   
29          0            NaN         63   среднее             1   
41          0            NaN         50   среднее             1   
55          0            NaN         54   среднее             1   
65          0            NaN         21   среднее             1   
67          0            NaN         52    высшее             0   
72          1            NaN         32    высшее             0   
82          2            NaN         50    высшее             0   
83          0            NaN         52   среднее             1   
90          2            NaN         35    высшее             0   
94          1            NaN         34    высшее             0   
96          0            NaN         44   СРЕДНЕЕ             1   
97          0            NaN         47    высшее  

Наблюдения:
1) Нет видимой связи пропущенных данных с остальными значениями.
2) Таких значений около 10%.

Проверим количество задолженностей

In [ ]:
print(null_data['debt'].sum())
print(null_data['debt'].count())
print(null_data['debt'].count() / null_data['debt'].sum())

170
2174
12.788235294117648


Наблюдения:
1) Процент невозвратов несколько выше, чем в среднем по другим группам

*Ответ:*  Гипотетическими причинами отсутствия связанных значений "доход" и "стаж" могут быть:
1) блокировка по критерию "кредитоспособность" - группа с высоким невозвратом;
2) отсутствие одного из значений автоматически анулирует второе;
3) при аккумулировании данных из разных источников произошёл сбой

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:*
1) Медиана представляет среднее значение в наборе данных.
2) Медиана важна, потому что она дает нам представление о том, где находится центральное значение в наборе данных.
3) Медиана имеет тенденцию быть более полезной для расчета, чем среднее значение, когда распределение асимметрично и/или имеет выбросы.


<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Комментарий ревьюера  ✔️</h2>

Совершенно верно, медиана не так чувствительна к выборосам по сравнению со средним, например. Получается, мы всегда можем брать медиану (кроме тех случаев, когда нужно именно среднее). Если выборка смещена, то мы не ошибемся, а если распределение нормальное, где среднее очень близко к медиане, то тоже не ошибемся. Только тут важно понимать, что не стоит проводить заполнение пропусков, взятой медианой по всему столбцу, потому что это тоже искажение данных.
</div>


### Шаг 4: общий вывод.

Напишите ваш общий вывод.

1) Уровень доходов не определяет уровень расходов. Задолженности по кредитам не имеют прямой зависимости от уровня дохода.  
2) Уровень расходов коррелируется с социальным классом, что наблюдается в исследованиях выше и оказывает влияние на наличие задолженностей.  
3) Коллективная ответственность помогает снижать задолженности.  
4) Задолженности по кредитам коррелируются с пирамидой Абрахама Маслоу, выражающей потребности заёмщиков.  
5) Данный набор является отражением кредитной политики банка и имеет тенденцию к предопределенности.  


В идеале надо приводить проценты невозврата в срок у самых надежных и самых безответственных клиентов для каждой группы. Например: самые надежные клиенты это люди, берущие кредит на операции с недвижимостью (такой-то процент невозврата в срок), а самые безответственные те, кто обращается в банк с целью получить деньги на покупку автомобиля (такой-то процент). И так для каждой группы. Это хороший тон, — подкреплять вывод полученными ранее результатами. Кроме того, так ты не оставляешь заказчику шансов неверно интерпретировать вывод.
</div>

1) Уровень доходов не определяет уровень расходов. Задолженности по кредитам не имеют прямой зависимости от уровня дохода:
    
    - Больше всего 15921 берут кредитов с уровнем дохода C: 50001 <= income <= 200000 - задолженности (8.50%)  
    - Далее 5014 берут кредиты с уровнем дохода B: 200001 <= income <= 1000000 - задолженности (7,06%)  
    - Далее 349 берут кредиты с уровнем дохода D: 30001 <= income <= 50000 - задолженности (6,02%)  
    - По доходу А: income >= 1000001 (не берут)  и Е: 0 <= income <= 30000 (не дают) - непрезентативны  

Прямой зависимости не наблюдается.
    
2) Уровень расходов коррелируется с социальным классом, что наблюдается в исследованиях выше и оказывает влияние на наличие задолженностей:
    
    - Группа вдов/вдовцов (6,62%)
    - Группа в разводах (7,06%)
    - Группа семейных (7,56%)  
    - Группа живущих в гражданском браке (9,31%)
    - Группа несемейных (9,76%)
    
Коллективная ответственность помогает снижать задолженности.
    
3) Задолженности по кредитам коррелируются с пирамидой Абрахама Маслоу, выражающей потребности заёмщиков:

    - Меньше всего задолженностей в группе с недвижимостью (7,26) - первая в пирамиде Маслоу. Самая большая ответственность;
    - Проведение свадьбы (7,91%) - групповое поручительство;
    - Операции с автомобилем (9,35%) и образованием (9,25%) - менее ответственный подход при взятии кредита.
    
4) Данный набор является отражением кредитной политики банка и имеет тенденцию к предопределенности.  

</div>